stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    then [uniq eq xs] has no duplicate elements at all.
+
+    [uniq eq xs] is equivalent to [map head (group eq xs)].
+
+    @since 4.14 *)
+
+val group : ('a -> 'a -> bool) -> 'a t -> 'a t t
+(** Provided the function [eq] defines an equality on elements,
+    [group eq xs] is the sequence of the maximal runs
+    of adjacent duplicate elements of the sequence [xs].
+
+    Every element of [group eq xs] is a nonempty sequence of equal elements.
+
+    The concatenation [concat (group eq xs)] is equal to [xs].
+
+    Consuming [group eq xs], and consuming the sequences that it contains,
+    can cause [xs] to be consumed more than once. Therefore, [xs] must be
+    persistent.
+
+    @since 4.14 *)
+
+val memoize : 'a t -> 'a t
+(** The sequence [memoize xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [memoize xs] is persistent: even if it is queried several times,
+    [xs] is queried at most once.
+    @since 4.14 *)
+
+exception ForcedTwice
+(** This exception is raised when a sequence returned by [once]
+    (or a suffix of it) is queried more than once.
+    @since 4.14 *)
+
+val once: 'a t -> 'a t
+(** The sequence [once xs] has the same elements as the sequence [xs].
+
+    Regardless of whether [xs] is ephemeral or persistent,
+    [once xs] is an ephemeral sequence: it can be queried at most once.
+    If it (or a suffix of it) is queried more than once, then the exception
+    [ForcedTwice] is raised. This can be useful, while debugging or testing,
+    to ensure that a sequence is consumed at most once.
+
+    @raise ForcedTwice if [once xs], or a suffix of it,
+           is queried more than once.
+    @since 4.14 *)
+
+val transpose : 'a t t -> 'a t t
+(** If [xss] is a matrix (a sequence of rows), then [transpose xss] is
+    the sequence of the columns of the matrix [xss].
+
+    The rows of the matrix [xss] are not required to have the same length.
+
+    The matrix [xss] is not required to be finite (in either direction).
+
+    The matrix [xss] must be persistent.
+
+    @since 4.14 *)
+
+val diagonals : 'a t t -> 'a t t
+(** If [xss] is a matrix (a sequence of rows), then [diagonals xss] is
+    the sequence of its diagonals.
+
+    The first diagonal contains just the first element of the
+    first row. The second diagonal contains the first element of the
+    second row and the second element of the first row; and so on.
+
+    Every diagonal is a finite sequence.
+
+    The rows of the matrix [xss] are not required to have the same length.
+
+    The matrix [xss] is not required to be finite (in either direction).
+
+    The matrix [xss] must be persistent.
+
+    @since 4.14 *)
+
+(** {1 Combining sequences} *)
+
+val append : 'a t -> 'a t -> 'a t
+(** [append xs ys] is the concatenation of the sequences [xs] and [ys].
+
+    Its elements are the elements of [xs], followed with the elements of [ys].
+
+    @since 4.11 *)
 
 val concat : 'a t t -> 'a t
```

Comment:
Note: it's not too hard to get the `?sep` stuff using `interleave` -- but it's noisier of course, so maybe a `concat_sep` would be in order.

Reply:
I thought about it before writing my comment but I came to the conclusion it would be non-trivial and I still believe so. Show me the light, what is the rune ?

Reply:
There is a proposal to add a function like `String.concat` to `List` [here](https://github.com/ocaml/ocaml/pull/9962), and to call it `List.intercalate` – in line with Haskell. Perhaps a similar case can be made here.